clang/lib/Sema/SemaCast.cpp

```diff
@@ -2568,6 +2570,16 @@ static TryCastResult TryAddressSpaceCast(Sema &Self, ExprResult &SrcExpr,
     msg = diag::err_bad_cxx_cast_addr_space_mismatch;
     return TC_Failed;
   }
+  if (Self.getLangOpts().SYCLIsDevice) {
+    Qualifiers SrcQ = SrcPointeeType.getQualifiers();
+    Qualifiers DestQ = DestPointeeType.getQualifiers();
+    if (SrcQ.isAddressSpaceSupersetOf(DestQ) &&
+        !DestQ.isAddressSpaceSupersetOf(SrcQ) && OpRange.isValid()) {
```

Comment:
Could you explain this logic please?

Reply:
All named address spaces (global, local, private) are disjoint and they are sub-sets of default address space. `isAddressSpaceSupersetOf` function answers the question if the address space of qualifiers is a superset of argument qualifiers.  We are trying to diagnose casts from default address space to named, in other words we need to diagnose when the cast happens from a big set to a small one. This check detects the case when `Src` pointer has address space that is a superset of `Dst` pointer's address space and at the same time address space of `Dst` pointer is not a superset of address space of `Src` pointer.
Also there is a verification that OpRange is valid. `TryAddressSpaceCast` function is also used for checks that are specifict for  C++ for OpenCL and not called for SYCL, so I didn't pass the `OpRange` argument there, since it is not needed there anyway. But added the check to make sure that the emitted warning always has a valid source location.

Reply:
> ... in other words we need to diagnose when the cast happens from a big set to a small one. This check detects the case when `Src` pointer has address space that is a superset of `Dst` pointer's address space and at the same time address space of `Dst` pointer is not a superset of address space of `Src` pointer.

Hi @Fznamznon, thank you, but this is the part that I would like to understand better.

Why isn't is enough to just check that 'Dst' is not a superset of 'Src'?   In other words, why is 'Src' being a superset of 'Dst' checked?

I understand the OpRange part.

Reply:
> Why isn't is enough to just check that 'Dst' is not a superset of 'Src'? In other words, why is 'Src' being a superset of 'Dst' checked?

Maybe it is just me being paranoid. I've changed it to check only that  'Dst' is not a superset of 'Src'. Thanks.